pkg/plugins/policies/meshtrafficpermission/graph/reachable_services_graph.go

```diff
@@ -0,0 +1,161 @@
+	if rule == nil {
+		return false
+	}
+	action := rule.Conf.(mtp_api.Conf).Action
+	return action == mtp_api.Allow || action == mtp_api.AllowWithShadowDeny
+}
+
+func Builder(resources context.Resources) context.ReachableServicesGraph {
+	services := BuildServices(resources.Dataplanes().Items, resources.ExternalServices().Items)
+	mtps := resources.ListOrEmpty(mtp_api.MeshTrafficPermissionType).(*mtp_api.MeshTrafficPermissionResourceList)
+	return BuildGraph(services, mtps.Items)
+}
+
+func BuildServices(
+	dataplanes []*mesh.DataplaneResource,
+	externalServices []*mesh.ExternalServiceResource,
+) map[string]mesh_proto.SingleValueTagSet {
+	services := map[string]mesh_proto.SingleValueTagSet{}
+	for _, dp := range dataplanes {
+		set := dp.Spec.TagSet()
+		for _, svc := range set.Values(mesh_proto.ServiceTag) {
+			if _, ok := services[svc]; ok {
+				continue
+			}
+			services[svc] = map[string]string{}
+			for tag := range SupportedTags {
+				if values := set.Values(tag); len(values) > 0 {
+					services[svc][tag] = values[0]
+				}
+			}
+		}
+	}
+	for _, es := range externalServices {
+		services[es.Spec.Tags[mesh_proto.ServiceTag]] = map[string]string{}
+	}
+	return services
+}
+
+func BuildGraph(services map[string]mesh_proto.SingleValueTagSet, mtps []*mtp_api.MeshTrafficPermissionResource) *Graph {
+	resources := context.Resources{
+		MeshLocalResources: map[core_model.ResourceType]core_model.ResourceList{
+			mtp_api.MeshTrafficPermissionType: &mtp_api.MeshTrafficPermissionResourceList{
+				Items: replaceTopLevelSubsets(mtps),
+			},
+		},
+	}
+
+	graph := NewGraph()
+
+	for service, tags := range services {
+		// build artificial dpp for matching
+		dp := mesh.NewDataplaneResource()
+		dpTags := maps.Clone(tags)
+		dpTags[mesh_proto.ServiceTag] = service
+		dp.Spec = &mesh_proto.Dataplane{
+			Networking: &mesh_proto.Dataplane_Networking{
+				Address: "1.1.1.1",
+				Inbound: []*mesh_proto.Dataplane_Networking_Inbound{
+					{
+						Tags: dpTags,
+						Port: 1234,
+					},
+				},
+			},
+		}
+
+		matched, err := matchers.MatchedPolicies(mtp_api.MeshTrafficPermissionType, dp, resources)
+		if err != nil {
+			log.Error(err, "service could not be matched. It won't be reached by any other service", "service", service)
+			continue // it's better to ignore one service that to break the whole graph
+		}
+
+		rl, ok := matched.FromRules.Rules[core_rules.InboundListener{
+			Address: "1.1.1.1",
+			Port:    1234,
+		}]
+		if !ok {
+			continue
+		}
+
+		graph.rules[service] = rl
+	}
+
+	return graph
+}
+
+// replaceTopLevelSubsets replaces subsets present in top-level target ref.
+// Because we need to do policy matching on services instead of individual proxies, we have to handle subsets in a special way.
+// What we do is we only support subsets with predefined tags listed in SupportedTags.
+// This assumes that tags listed in SupportedTags are stable between all instances of a given service.
```

Comment:
What does it mean "stable between all instances of a given service"? You can have 1 service spread across many namespaces right?

Reply:
stable = the same. All instances of a service has the same value of a tag